Task `help --task help` fails on the second run when configuration cache is enabled

### Expected Behavior
The task runs with the same output.

### Current Behavior
The task runs normally on the first run and fails on the second one with an error:
```
Rustam.Musin@UNIT-2257 gradle-help-task-issue % ./gradlew help --task help
Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: help --task help

> Task :help
Detailed task information for help

Path
     :help

Type
     Help (org.gradle.configuration.Help)

Options
     --task     The task to show help for.

Description
     Displays a help message.

Group
     help

0 problems were found storing the configuration cache.

See the complete report at file:///Users/Rustam.Musin/my/gradle-help-task-issue/build/reports/configuration-cache/9mvlkblqk1hawnen5cx3q7g1e/b3tsmfa1prg1pxaqtnzmgrazr/configuration-cache-report.html

BUILD SUCCESSFUL in 744ms
1 actionable task: 1 executed
Configuration cache entry stored.
Rustam.Musin@UNIT-2257 gradle-help-task-issue % ./gradlew help --task help
Configuration cache is an incubating feature.
Reusing configuration cache.
> Task :help FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':help'.
> A problem occurred configuring root project 'gradle-help-task-issue'.
   > Failed to apply plugin 'org.gradle.help-tasks'.
      > Cannot add task 'help' as a task with that name already exists.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 543ms
1 actionable task: 1 executed
Configuration cache entry reused.
Rustam.Musin@UNIT-2257 gradle-help-task-issue % 
```

### Context
Just wanted to try out the configuration cache, and the attempt has been failed :(

### Steps to Reproduce 
0. Use Gradle 7.4.2 on macOS Monterey x86_64
1. Download the project
2. Run `./gradlew help --task help` and see the correct output
3. Run the same command again and see the error

I've attached the sample project
[gradle-help-task-issue.zip](https://github.com/gradle/gradle/files/8948630/gradle-help-task-issue.zip).

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIntegrationTest.groovy
@@ -25,22 +25,74 @@ import spock.lang.Issue
 
 class ConfigurationCacheIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
 
-    def "configuration cache for help on empty project"() {
+    def "configuration cache for Help plugin task '#task' on empty project"() {
         given:
         settingsFile.createFile()
-        configurationCacheRun "help"
+        configurationCacheRun(task, *options)
         def firstRunOutput = removeVfsLogOutput(result.normalizedOutput)
-            .replaceAll(/Calculating task graph as no configuration cache is available for tasks: help\n/, '')
+            .replaceAll(/Calculating task graph as no configuration cache is available for tasks: ${task}.*\n/, '')
             .replaceAll(/Configuration cache entry stored.\n/, '')
 
         when:
-        configurationCacheRun "help"
+        configurationCacheRun(task, *options)
         def secondRunOutput = removeVfsLogOutput(result.normalizedOutput)
             .replaceAll(/Reusing configuration cache.\n/, '')
             .replaceAll(/Configuration cache entry reused.\n/, '')
 
         then:
         firstRunOutput == secondRunOutput
+
+        where:
+        task            | options
+        "help"          | []
+        "properties"    | []
+        "dependencies"  | []
+        "help"          | ["--task", "help"]
+    }
+
+    def "can store task selection success/failure for :help --task"() {
+        def configurationCache = newConfigurationCacheFixture()
+        buildFile.text = """
+        task aTask
+        """
+        when:
+        configurationCacheFails "help", "--task", "bTask"
+        then:
+        failure.assertHasCause("Task 'bTask' not found in root project")
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheFails "help", "--task", "cTask"
+        then:
+        failure.assertHasCause("Task 'cTask' not found in root project")
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheFails "help", "--task", "bTask"
+        then:
+        failure.assertHasCause("Task 'bTask' not found in root project")
+        configurationCache.assertStateLoaded()
+
+        when:
+        configurationCacheRun "help", "--task", "aTask"
+        then:
+        output.contains "Detailed task information for aTask"
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheFails "help", "--task", "cTask"
+        then:
+        failure.assertHasCause("Task 'cTask' not found in root project")
+        configurationCache.assertStateLoaded()
+
+        when:
+        buildFile << """
+        task bTask
+        """
+        configurationCacheRun "help", "--task", "bTask"
+        then:
+        output.contains "Detailed task information for bTask"
+        configurationCache.assertStateStored()
     }
 
     @Issue("https://github.com/gradle/gradle/issues/18064")

--- a/subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/HelpTaskIntegrationTest.groovy
+++ b/subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/HelpTaskIntegrationTest.groovy
@@ -461,11 +461,29 @@ BUILD SUCCESSFUL"""
     def "error message contains possible candidates"() {
         buildFile.text = """
         task aTask
-"""
+        """
         when:
         fails "help", "--task", "bTask"
         then:
         failure.assertHasCause("Task 'bTask' not found in root project '${testDirectory.getName()}'. Some candidates are: 'aTask', 'tasks'")
+
+        when:
+        run "help", "--task", "aTask"
+        then:
+        output.contains "Detailed task information for aTask"
+
+        when:
+        fails "help", "--task", "bTask"
+        then:
+        failure.assertHasCause("Task 'bTask' not found in root project '${testDirectory.getName()}'. Some candidates are: 'aTask', 'tasks'")
+
+        when:
+        buildFile << """
+        task bTask
+        """
+        run "help", "--task", "bTask"
+        then:
+        output.contains "Detailed task information for bTask"
     }
 
     def "tasks can be defined by camelCase matching"() {

--- a/subprojects/diagnostics/src/main/java/org/gradle/configuration/Help.java
+++ b/subprojects/diagnostics/src/main/java/org/gradle/configuration/Help.java
@@ -18,9 +18,10 @@ package org.gradle.configuration;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.internal.DocumentationRegistry;
 import org.gradle.api.internal.tasks.options.OptionReader;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Property;
 import org.gradle.api.tasks.TaskAction;
 import org.gradle.api.tasks.options.Option;
-import org.gradle.execution.TaskSelection;
 import org.gradle.execution.TaskSelector;
 import org.gradle.initialization.BuildClientMetaData;
 import org.gradle.initialization.layout.ResolvedBuildLayout;
@@ -35,7 +36,23 @@ import static org.gradle.internal.logging.text.StyledTextOutput.Style.UserInput;
 
 @DisableCachingByDefault(because = "Produces only non-cacheable console output")
 public class Help extends DefaultTask {
-    private String taskPath;
+    private final Property<String> taskPath = getObjectFactory().property(String.class);
+
+    private final Property<TaskDetailsModel> taskModel = getObjectFactory().property(TaskDetailsModel.class).convention(taskPath.map(this::mapFromTaskPath));
+
+    public Help() {
+        // optimization: so value does not need to be recomputed during execution
+        taskModel.finalizeValueOnRead();
+    }
+
+    private Property<String> getTaskPath() {
+        return taskPath;
+    }
+
+    @Inject
+    protected ObjectFactory getObjectFactory() {
+        throw new UnsupportedOperationException();
+    }
 
     @Inject
     protected StyledTextOutputFactory getTextOutputFactory() {
@@ -71,7 +88,7 @@ public class Help extends DefaultTask {
     void displayHelp() {
         StyledTextOutput output = getTextOutputFactory().create(Help.class);
         BuildClientMetaData metaData = getClientMetaData();
-        if (taskPath != null) {
+        if (getTaskPath().isPresent()) {
             printTaskHelp(output);
         } else {
             printDefaultHelp(output, metaData);
@@ -79,10 +96,8 @@ public class Help extends DefaultTask {
     }
 
     private void printTaskHelp(StyledTextOutput output) {
-        TaskSelector selector = getTaskSelector();
-        TaskSelection selection = selector.getSelection(taskPath);
-        OptionReader optionReader = getOptionReader();
-        TaskDetailPrinter taskDetailPrinter = new TaskDetailPrinter(taskPath, selection, optionReader);
+        TaskDetailsModel taskDetailModel = taskModel.get();
+        TaskDetailPrinter taskDetailPrinter = new TaskDetailPrinter(taskDetailModel.getTaskPath(), taskDetailModel.getTasks());
         taskDetailPrinter.print(output);
     }
 
@@ -136,6 +151,10 @@ public class Help extends DefaultTask {
 
     @Option(option = "task", description = "The task to show help for.")
     public void setTaskPath(String taskPath) {
-        this.taskPath = taskPath;
+        this.getTaskPath().set(taskPath);
+    }
+
+    private TaskDetailsModel mapFromTaskPath(String taskPath) {
+        return TaskDetailsModel.from(taskPath, getTaskSelector(), getOptionReader());
     }
 }

--- a/subprojects/diagnostics/src/main/java/org/gradle/configuration/TaskDetailPrinter.java
+++ b/subprojects/diagnostics/src/main/java/org/gradle/configuration/TaskDetailPrinter.java
@@ -15,24 +15,15 @@
  */
 package org.gradle.configuration;
 
-import com.google.common.collect.ArrayListMultimap;
-import com.google.common.collect.ListMultimap;
 import com.google.common.collect.Sets;
 import org.apache.commons.lang.StringUtils;
-import org.gradle.api.DefaultTask;
 import org.gradle.api.NonNullApi;
-import org.gradle.api.Task;
 import org.gradle.api.Transformer;
-import org.gradle.api.internal.plugins.DslObject;
-import org.gradle.api.internal.tasks.options.OptionDescriptor;
-import org.gradle.api.internal.tasks.options.OptionReader;
-import org.gradle.execution.TaskSelection;
 import org.gradle.internal.logging.text.LinePrefixingStyledTextOutput;
 import org.gradle.internal.logging.text.StyledTextOutput;
 
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -41,48 +32,46 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.TreeSet;
+import java.util.stream.Collectors;
 
 import static org.gradle.internal.logging.text.StyledTextOutput.Style.UserInput;
 import static org.gradle.util.internal.CollectionUtils.collect;
-import static org.gradle.util.internal.CollectionUtils.filter;
 import static org.gradle.util.internal.CollectionUtils.sort;
 
 @NonNullApi
 public class TaskDetailPrinter {
     private final String taskPath;
-    private final TaskSelection selection;
+    private final List<TaskDetails> tasks;
     private static final String INDENT = "     ";
-    private final OptionReader optionReader;
 
-    public TaskDetailPrinter(String taskPath, TaskSelection selection, OptionReader optionReader) {
+    public TaskDetailPrinter(String taskPath, List<TaskDetails> tasks) {
         this.taskPath = taskPath;
-        this.selection = selection;
-        this.optionReader = optionReader;
+        this.tasks = tasks;
     }
 
     public void print(StyledTextOutput output) {
-        final List<Task> tasks = sort(selection.getTasks());
-
         output.text("Detailed task information for ").withStyle(UserInput).println(taskPath);
-        final ListMultimap<Class<?>, Task> classListMap = groupTasksByType(tasks);
+        final Map<String, List<TaskDetails>> detailsByTaskType = groupTasksByType(this.tasks);
 
-        final Set<Class<?>> classes = classListMap.keySet();
-        boolean multipleClasses = classes.size() > 1;
-        final List<Class<?>> sortedClasses = sort(classes, Comparator.comparing(Class::getSimpleName));
-        for (Class<?> clazz : sortedClasses) {
+        final Set<String> typeNames = detailsByTaskType.keySet();
+        boolean multipleTaskTypes = typeNames.size() > 1;
+        final List<String> sortedTaskTypes = sort(typeNames);
+        for (String taskType : sortedTaskTypes) {
             output.println();
-            final List<Task> tasksByType = classListMap.get(clazz);
+            final List<TaskDetails> tasksByType = detailsByTaskType.get(taskType);
+            final TaskDetails anyTask = tasksByType.iterator().next();
+            String shortTypeName = anyTask.getShortTypeName();
             final LinePrefixingStyledTextOutput pathOutput = createIndentedOutput(output, INDENT);
             pathOutput.println(tasksByType.size() > 1 ? "Paths" : "Path");
-            for (Task task : tasksByType) {
+            for (TaskDetails task : tasksByType) {
                 pathOutput.withStyle(UserInput).println(task.getPath());
             }
 
             output.println();
             final LinePrefixingStyledTextOutput typeOutput = createIndentedOutput(output, INDENT);
             typeOutput.println("Type");
-            typeOutput.withStyle(UserInput).text(clazz.getSimpleName());
-            typeOutput.println(" (" + clazz.getName() + ")");
+            typeOutput.withStyle(UserInput).text(shortTypeName);
+            typeOutput.println(" (" + taskType + ")");
 
             printlnCommandlineOptions(output, tasksByType);
 
@@ -92,53 +81,37 @@ public class TaskDetailPrinter {
             output.println();
             printTaskGroup(output, tasksByType);
 
-            if (multipleClasses) {
+            if (multipleTaskTypes) {
                 output.println();
                 output.println("----------------------");
             }
         }
     }
 
-    private ListMultimap<Class<?>, Task> groupTasksByType(List<Task> tasks) {
-        final Set<Class<?>> taskTypes = new TreeSet<>(Comparator.comparing(Class::getSimpleName));
-        taskTypes.addAll(collect(tasks, this::getDeclaredTaskType));
-
-        ListMultimap<Class<?>, Task> tasksGroupedByType = ArrayListMultimap.create();
-        for (final Class<?> taskType : taskTypes) {
-            tasksGroupedByType.putAll(taskType, filter(tasks, element -> getDeclaredTaskType(element).equals(taskType)));
-        }
-        return tasksGroupedByType;
+    private Map<String, List<TaskDetails>> groupTasksByType(Collection<TaskDetails> tasks) {
+        return tasks.stream().collect(Collectors.groupingBy(TaskDetails::getTaskType, LinkedHashMap::new, Collectors.toList()));
     }
 
-    private Class<?> getDeclaredTaskType(Task original) {
-        Class<?> clazz = new DslObject(original).getDeclaredType();
-        if (clazz.equals(DefaultTask.class)) {
-            return org.gradle.api.Task.class;
-        } else {
-            return clazz;
-        }
+    private void printTaskDescription(StyledTextOutput output, List<TaskDetails> tasks) {
+        printTaskAttribute(output, "Description", tasks, TaskDetails::getDescription);
     }
 
-    private void printTaskDescription(StyledTextOutput output, List<Task> tasks) {
-        printTaskAttribute(output, "Description", tasks, Task::getDescription);
+    private void printTaskGroup(StyledTextOutput output, List<TaskDetails> tasks) {
+        printTaskAttribute(output, "Group", tasks, TaskDetails::getGroup);
     }
 
-    private void printTaskGroup(StyledTextOutput output, List<Task> tasks) {
-        printTaskAttribute(output, "Group", tasks, Task::getGroup);
-    }
-
-    private void printTaskAttribute(StyledTextOutput output, String attributeHeader, List<Task> tasks, Transformer<String, Task> transformer) {
+    private void printTaskAttribute(StyledTextOutput output, String attributeHeader, List<TaskDetails> tasks, Transformer<String, TaskDetails> transformer) {
         int count = collect(tasks, new HashSet<>(), transformer).size();
         final LinePrefixingStyledTextOutput attributeOutput = createIndentedOutput(output, INDENT);
         if (count == 1) {
             // all tasks have the same value
             attributeOutput.println(attributeHeader);
-            final Task task = tasks.iterator().next();
+            final TaskDetails task = tasks.iterator().next();
             String value = transformer.transform(task);
             attributeOutput.println(value == null ? "-" : value);
         } else {
             attributeOutput.println(attributeHeader + "s");
-            for (Task task : tasks) {
+            for (TaskDetails task : tasks) {
                 attributeOutput.withStyle(UserInput).text("(" + task.getPath() + ") ");
                 String value = transformer.transform(task);
                 attributeOutput.println(value == null ? "-" : value);
@@ -146,10 +119,10 @@ public class TaskDetailPrinter {
         }
     }
 
-    private void printlnCommandlineOptions(StyledTextOutput output, List<Task> tasks) {
-        List<OptionDescriptor> allOptions = new ArrayList<>();
-        for (Task task : tasks) {
-            allOptions.addAll(optionReader.getOptions(task));
+    private void printlnCommandlineOptions(StyledTextOutput output, List<TaskDetails> tasks) {
+        List<TaskDetails.OptionDetails> allOptions = new ArrayList<>();
+        for (TaskDetails task : tasks) {
+            allOptions.addAll(task.getOptions());
         }
         if (!allOptions.isEmpty()) {
             output.println();
@@ -182,9 +155,9 @@ public class TaskDetailPrinter {
         }
     }
 
-    private Map<String, Set<String>> optionToAvailableValues(List<OptionDescriptor> allOptions) {
+    private Map<String, Set<String>> optionToAvailableValues(List<TaskDetails.OptionDetails> allOptions) {
         Map<String, Set<String>> result = new LinkedHashMap<>();
-        for (OptionDescriptor optionDescriptor : allOptions) {
+        for (TaskDetails.OptionDetails optionDescriptor : allOptions) {
             if (result.containsKey(optionDescriptor.getName())) {
                 Collection<String> commonValues = Sets.intersection(optionDescriptor.getAvailableValues(), result.get(optionDescriptor.getName()));
                 result.put(optionDescriptor.getName(), new TreeSet<>(commonValues));
@@ -195,9 +168,9 @@ public class TaskDetailPrinter {
         return result;
     }
 
-    private Map<String, String> optionToDescription(List<OptionDescriptor> allOptions) {
+    private Map<String, String> optionToDescription(List<TaskDetails.OptionDetails> allOptions) {
         Map<String, String> result = new HashMap<>();
-        for (OptionDescriptor optionDescriptor : allOptions) {
+        for (TaskDetails.OptionDetails optionDescriptor : allOptions) {
             result.put(optionDescriptor.getName(), optionDescriptor.getDescription());
         }
         return result;

--- a/subprojects/diagnostics/src/main/java/org/gradle/configuration/TaskDetails.java
+++ b/subprojects/diagnostics/src/main/java/org/gradle/configuration/TaskDetails.java
@@ -1,0 +1,156 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configuration;
+
+import org.apache.commons.lang.StringUtils;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.NonNullApi;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.internal.plugins.DslObject;
+import org.gradle.api.internal.tasks.options.OptionDescriptor;
+import org.gradle.api.internal.tasks.options.OptionReader;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A configuration-cache friendly view of @{link Task} that only projects information
+ * that is relevant for the @{link Help#} task.
+ */
+@NonNullApi
+class TaskDetails {
+    final static Comparator<TaskDetails> DEFAULT_COMPARATOR = (o1, o2) -> {
+        // tasks in higher-up projects first
+        int depthCompare = o1.getProjectDepth() - o2.getProjectDepth();
+        if (depthCompare != 0) {
+            return depthCompare;
+        }
+        return o1.getPath().compareTo(o2.getPath());
+    };
+
+    /**
+     * A read-only projection for OptionDescriptor details relevant for the help task.
+     */
+    public static class OptionDetails {
+        private final String name;
+
+        private final String description;
+
+        private final Set<String> availableValues;
+        public OptionDetails(String name, String description, Set<String> availableValues) {
+            this.name = name;
+            this.description = description;
+            this.availableValues = availableValues;
+        }
+
+        private static OptionDetails from(OptionDescriptor option) {
+            return new OptionDetails(option.getName(), option.getDescription(), option.getAvailableValues());
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public String getDescription() {
+            return description;
+        }
+
+        public Set<String> getAvailableValues() {
+            return availableValues;
+        }
+    }
+
+    private final String path;
+
+    private final String taskType;
+
+    private final String shortTypeName;
+
+    private final String description;
+
+    private final String group;
+
+    private final int projectDepth;
+
+    private final List<OptionDetails> options;
+
+    private TaskDetails(String path, String taskType, String shortTypeName, @Nullable String description, @Nullable String group, int projectDepth, List<OptionDetails> options) {
+        this.path = path;
+        this.taskType = taskType;
+        this.shortTypeName = shortTypeName;
+        this.description = description;
+        this.group = group;
+        this.projectDepth = projectDepth;
+        this.options = options;
+    }
+
+    /**
+     * The task type implementation Java class.
+     */
+    public String getTaskType() {
+        return taskType;
+    }
+
+    public String getShortTypeName() {
+        return shortTypeName;
+    }
+
+    @Nullable
+    public String getDescription() {
+        return description;
+    }
+
+    @Nullable
+    public String getGroup() {
+        return group;
+    }
+
+    public String getPath() {
+        return path;
+    }
+
+    public int getProjectDepth() {
+        return projectDepth;
+    }
+
+    public List<OptionDetails> getOptions() {
+        return options;
+    }
+
+    public static TaskDetails from(Task task, OptionReader optionReader) {
+        String path = task.getPath();
+        int projectDepth = StringUtils.countMatches(path, Project.PATH_SEPARATOR);
+        List<OptionDetails> options = optionReader.getOptions(task).stream().map(OptionDetails::from).collect(Collectors.toList());
+        Class<?> declaredTaskType = getDeclaredTaskType(task);
+        String taskType = declaredTaskType.getName();
+        String shortTypeName = declaredTaskType.getSimpleName();
+        return new TaskDetails(path, taskType, shortTypeName, task.getDescription(), task.getGroup(), projectDepth, options);
+    }
+
+    private static Class<?> getDeclaredTaskType(Task original) {
+        Class<?> clazz = new DslObject(original).getDeclaredType();
+        if (clazz.equals(DefaultTask.class)) {
+            return org.gradle.api.Task.class;
+        } else {
+            return clazz;
+        }
+    }
+}

--- a/subprojects/diagnostics/src/main/java/org/gradle/configuration/TaskDetailsModel.java
+++ b/subprojects/diagnostics/src/main/java/org/gradle/configuration/TaskDetailsModel.java
@@ -1,0 +1,71 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configuration;
+
+import org.gradle.api.Task;
+import org.gradle.api.internal.tasks.options.OptionReader;
+import org.gradle.execution.TaskSelectionException;
+import org.gradle.execution.TaskSelector;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * A configuration-friendly view of a task selection.
+ */
+class TaskDetailsModel {
+    private final String taskPath;
+    private final List<TaskDetails> tasks;
+    private final TaskSelectionException failure;
+
+    private TaskDetailsModel(String taskPath, List<TaskDetails> tasks) {
+        this.taskPath = taskPath;
+        this.failure = null;
+        this.tasks = tasks;
+    }
+
+    private TaskDetailsModel(String taskPath, TaskSelectionException failure) {
+        this.taskPath = taskPath;
+        this.tasks = Collections.emptyList();
+        this.failure = failure;
+    }
+    public List<TaskDetails> getTasks() {
+        if (failure != null) {
+            // rethrow the original failure
+            throw failure;
+        }
+        return tasks;
+    }
+
+    public String getTaskPath() {
+        return taskPath;
+    }
+
+    public static TaskDetailsModel from(String taskPath, TaskSelector taskSelector, OptionReader optionReader) {
+        try {
+            Stream<Task> selectedTasks = taskSelector.getSelection(taskPath).getTasks().stream();
+            List<TaskDetails> tasks = selectedTasks.map(t -> TaskDetails.from(t, optionReader))
+                .sorted(TaskDetails.DEFAULT_COMPARATOR).collect(Collectors.toList());
+            return new TaskDetailsModel(taskPath, tasks);
+        } catch (TaskSelectionException exception) {
+            // collect exception so we can rethrow it during task execution
+            return new TaskDetailsModel(taskPath, exception);
+        }
+    }
+}

